### PR TITLE
Only store segmentID for flushed segments event

### PR DESCRIPTION
**Pull request by xiaocai2333**
Signed-off-by: cai.zhang <cai.zhang@zilliz.com>
issue: #18642

### CODE REVIEW

**Comment by xiaocai2333** [comment]
rerun ut

File: internal/indexcoord/index_builder.go [hunk 1]
```diff
@@ -118,19 +117,26 @@ func (ib *indexBuilder) reloadFromKV(aliveNodes []UniqueID) {
...
+
 	ib.taskMutex.Lock()
 	defer ib.taskMutex.Unlock()
+
 	ib.tasks[buildID] = indexTaskInit
-	select {
-	case ib.notify <- struct{}{}:
-	default:
-	}
+	ib.notify()
```
Comment (jiaoew1991) [inline internal/indexcoord/index_builder.go:135]: ib.notify() called twice?

File: internal/indexcoord/index_builder.go [hunk 2]
```diff
@@ -141,43 +147,81 @@ func (ib *indexBuilder) schedule() {
...
+		log.Debug("index task has not exist in meta table, remove task", zap.Int64("buildID", buildID))
+		delete(ib.tasks, buildID)
+		return
+	}
 
 	switch state {
 	case indexTaskInit:
-		if !exist || !ib.meta.NeedIndex(meta.CollectionID, meta.IndexID) {
+		if !ib.meta.NeedIndex(meta.CollectionID, meta.IndexID) {
 			delete(ib.tasks, buildID)
```
Comment (jiaoew1991) [inline internal/indexcoord/index_builder.go:206]: use deleteFunc ?

Comment (jiaoew1991) [reply internal/indexcoord/index_builder.go:206]: the same as L199

File: internal/indexcoord/flush_segment_watcher.go [hunk 3]
```diff
@@ -255,50 +260,38 @@ func (fsw *flushedSegmentWatcher) childrenRun() {
...
-		case fsw.internalNotify <- struct{}{}:
-		default:
-		}
-		return
+		fsw.constructTask(t)
+		fsw.internalTasks[segID].state = indexTaskInProgress
+		fsw.internalNotifyFunc()
 	case indexTaskInProgress:
 		fsw.childrenTaskMutex.RLock()
 		defer fsw.childrenTaskMutex.RUnlock()
```
Comment (czs007) [inline internal/indexcoord/flush_segment_watcher.go:287]: do not use defer here.
case indexTaskDeleted  alse has a defer.

File: internal/indexcoord/flush_segment_watcher.go [hunk 4]
```diff
@@ -229,13 +225,22 @@ func (fsw *flushedSegmentWatcher) childrenScheduler() {
 }
 
 func (fsw *flushedSegmentWatcher) internalRun() {
+	segmentIDs := make([]UniqueID, 0)
 	fsw.internalTaskMutex.Lock()
```
Comment (czs007) [inline internal/indexcoord/flush_segment_watcher.go:229]: this lock holds too long.
there are grpc calls in fsw.internalProcess

Comment (czs007) [reply internal/indexcoord/flush_segment_watcher.go:229]: If the lock is messy, you can consider to use sync.Map

File: internal/indexcoord/flush_segment_watcher.go [hunk 5]
```diff
@@ -313,25 +306,24 @@ func (fsw *flushedSegmentWatcher) internalProcess(t *internalTask) {
 		}
 		if err := fsw.writeHandoffSegment(handoffTask); err != nil {
 			log.Error("IndexCoord flushSegmentWatcher writeHandoffSegment with no index fail",
-				zap.Int64("segID", t.segmentInfo.ID), zap.Error(err))
+				zap.Int64("segID", segID), zap.Error(err))
 			return
 		}
 		if err := fsw.removeFlushedSegment(t); err != nil {
```
Comment (czs007) [inline internal/indexcoord/flush_segment_watcher.go:312]: Is it now allowed to delete CompactFrom immediately?

Comment (czs007) [reply internal/indexcoord/flush_segment_watcher.go:312]: SegmentA and SegmentB merged to SegmentC.
if Segment C created index, do we have right to delete Index of SegmentA and SegmentB ?

File: internal/indexcoord/flush_segment_watcher.go [hunk 6]
```diff
@@ -477,3 +476,61 @@ func (fsw *flushedSegmentWatcher) removeFlushedSegment(t *internalTask) error {
 		zap.Int64("partID", t.segmentInfo.PartitionID), zap.Int64("segID", t.segmentInfo.ID))
 	return nil
 }
+
+func (fsw *flushedSegmentWatcher) getSegmentInfo(segmentID UniqueID) error {
```
Comment (czs007) [inline internal/indexcoord/flush_segment_watcher.go:480]: suggest to rename as pullSegmentInfo

**Pull request merged**